lib,zebra: store resolving nexthop-group id with recursive nexthops

mjstapp opened the pull request:
When resolving recursive nexthops in zebra, capture the route prefix and id of the nhg used to resolve each nexthop. The goal of this is mainly to offer better support for FIB managers living in the NOS hosting FRR/zebra. Zebra forms fully-resolved  or "flattened" nexthop objects, and those objects contain contributions from both the original nexthop (sent by the owning protocol daemon) and from the resolving nexthop. A FIB manager can use the resolved-via id to distinguish the contributions from the various possible sources. That might be of use on a platform that supports a hierarchical FIB, for example. The sonic OS is expecting to take advantage of this information as its FIB manager evolves, we believe.

[7/29 I'm adding the "don't merge" label. I've added resolving prefix info to the nexthop struct, and that may be more useful than the id, but it's more data and we may not want to use that approach.]

Commit:
lib,zebra: track the id of the nhgs used to resolve nexthops

When resolving recursive nexthops, capture the id of the nhg
used to resolve each nexthop. This will allow us to identify
the contribution the resolution processing makes to the
final "flattened" resolved nexthop.

Signed-off-by: Mark Stapp <mjs@cisco.com>

Commit:
lib,zebra: add resolved-via id to show output

Add the resolved-via value to show output (vty and json
forms).

Signed-off-by: Mark Stapp <mjs@cisco.com>

Commit:
zebra: Encode resolved-via field for FPM

Add a number-space for FPM nexthop-group attribute encoding,
and add a value for the resolved-via field. Encode the
resolved-via field for FPM nexthop-group messages.

Signed-off-by: Mark Stapp <mjs@cisco.com>

ton31337 commented on zebra/kernel_netlink.c:
```diff
@@ -1564,8 +1564,18 @@ static enum netlink_msg_status nl_put_msg(struct nl_batch *bth,
+		 */
+		if (dplane_ctx_get_nhe_id(ctx) == dplane_ctx_get_old_nhe_id(ctx)) {
+			return FRR_NETLINK_SUCCESS;
+			break;
```
Dead code?

ton31337 commented on zebra/zebra_vty.c:
```diff
@@ -327,6 +327,9 @@ static void show_nexthop_detail_helper(struct vty *vty,
 	}
+
+	if (nexthop->resolved_via > 0)
+		vty_out(vty, ", res via %u", nexthop->resolved_via);
```
nit: can't we use a full word (resolved) here?

riw777 reviewed (dismissed):
looks good (other than Donatas' comments)

riw777 commented:
Looks like this lint needs to be fixed, too --

111< WARNING: break is not useful after a return
< #1573: FILE: /tmp/f1-167235/kernel_netlink.c:1573:
Details at [https://ci1.netdef.org/browse/FRR-PULLREQ3-CHECKOUT-10307](https://ci1.netdef.org/browse/FRR-PULLREQ3-CHECKOUT-10307)```

mjstapp commented:
rebased to newer master

mjstapp commented:
removed a commit about optimizing some kernel updates - not relevant for this specific feature

eddieruan-alibaba commented on zebra/zebra_nhg.c:
```diff
@@ -2674,6 +2674,12 @@ static int nexthop_active(struct nexthop *nexthop, struct nhg_hash_entry *nhe,
+				 */
+				if (resolver)
+					resolver->resolved_via = match->nhe_id;
+
```
Use the example I put in  https://github.com/FRRouting/frr/issues/22734

PE3# show nexthop-group rib 131
ID: 131 (zebra)
RefCnt: 2
Uptime: 00:03:27
VRF: default(IPv6)
Nexthop Count: 1
Flags: 0x4009
Valid, Installed (FPM only), Recursive
Depends: (109)
via 2064:200::1e (vrf default) (recursive), weight 1
via fc06::2, Ethernet12 (vrf default), weight 1
via fc08::2, Ethernet4 (vrf default), weight 1


PE3# show nexthop-group rib 136
ID: 136 (zebra)
RefCnt: 1
Uptime: 00:03:24
VRF: default(No AFI)
Nexthop Count: 2
Flags: 0x3
Valid, Installed
Depends: (131) (137)
via 2064:100::1d (vrf default) (recursive), weight 1
via fc06::2, Ethernet12 (vrf default), weight 1
via fc08::2, Ethernet4 (vrf default), weight 1
via 2064:200::1e (vrf default) (recursive), weight 1
via fc06::2, Ethernet12 (vrf default) (dup), weight 1
via fc08::2, Ethernet4 (vrf default) (dup), weight 1

My expectation is that

   NHG 131 (recursive singleton for 2064:200::1e, Depends: (109)):

   The route to 2064:200::1e is installed with NHE 109 (that's exactly why 131 depends on 109 — a recursive NHG's depend is the resolving route's NHE). So:
    via 2064:200::1e (recursive), weight 1          <- resolved_via = 0 (not shown)
       via fc06::2, Ethernet12, weight 1            <- resolved_via = 109
       via fc08::2, Ethernet4, weight 1             <- resolved_via = 109

   NHG 136 (group of 131 + 137):
   - 137 is the recursive singleton for 2064:100::1d. Its resolving route's nexthops are the identical pair (fc06::2/Ethernet12, fc08::2/Ethernet4), and since zebra shares NHEs by hash, that route's nhe_id is also 109.
   - 136's nexthop list is a deep copy of its depends' nexthops (resolved_via is preserved by the copy — the PR adds it to nexthop_copy_no_recurse()), so:
    via 2064:100::1d (recursive), weight 1          <- resolved_via = 0
       via fc06::2, Ethernet12, weight 1            <- resolved_via = 109
       via fc08::2, Ethernet4, weight 1             <- resolved_via = 109
    via 2064:200::1e (recursive), weight 1          <- resolved_via = 0
       via fc06::2, Ethernet12 (dup), weight 1      <- resolved_via = 109
       via fc08::2, Ethernet4 (dup), weight 1       <- resolved_via = 109


If the direct link goes away, we could use resolved_via to fix up 109. 

But if 2064:100::1d got withdrew, how do we update NHG 136 to remove  1d path, so we could buy time for zebra to resolve proper paths again. 


Based on today's discussion, I am looking at the binding information, a.k.a resolve via prefix. 

 2064:100::1d --> NHG 136
 2064:200::1e --> NHG 136

When route  2064:100::1d delete event comes to dplane, we could use this route to find out impacted NHGs and make quick fix up. This is for not using NHT event, but using route delete event to trigger fix up.   This is for https://github.com/FRRouting/frr/issues/22734. 

When route  2064:100::1d comes back before NHG 136 gets deleted (revive case), we could use this binding information to update NHG 136 properly. This is for https://github.com/FRRouting/frr/issues/22732

eddieruan-alibaba commented:
@mjstapp Put your diff into our testbed. We could get resolvedVia in json, but I don't see resolve via in regular "show nexthop rib <id>"  output


PE3# show nexthop rib 145 json
{
  "145":{
    "type":"zebra",
    "refCount":1,
    "uptime":"00:03:24",
    "vrf":"default",
    "afi":"No AFI",
    "nexthopCount":2,
    "flags":3,
    "valid":true,
    "installed":true,
    "depends":[
      142,
      146
    ],
    "nexthops":[
      {
        "flags":5,
        "ip":"2064:100::1d",
        "afi":"ipv6",
        "vrf":"default",
        "active":true,
        "recursive":true,
        "weight":1
      },
      {
        "flags":3,
        "fib":true,
        "ip":"fc06::2",
        "afi":"ipv6",
        "interfaceIndex":51,
        "interfaceName":"Ethernet12",
        "vrf":"default",
        "resolver":true,
        "active":true,
        "weight":1,
        "resolvedVia":117
      },
      {
        "flags":3,
        "fib":true,
        "ip":"fc08::2",
        "afi":"ipv6",
        "interfaceIndex":49,
        "interfaceName":"Ethernet4",
        "vrf":"default",
        "resolver":true,
        "active":true,
        "weight":1,
        "resolvedVia":117
      },
      {
        "flags":5,
        "ip":"2064:200::1e",
        "afi":"ipv6",
        "vrf":"default",
        "active":true,
        "recursive":true,
        "weight":1
      },
      {
        "flags":17,
        "duplicate":true,
        "ip":"fc06::2",
        "afi":"ipv6",
        "interfaceIndex":51,
        "interfaceName":"Ethernet12",
        "vrf":"default",
        "resolver":true,
        "active":true,
        "weight":1,
        "resolvedVia":117
      },
      {
        "flags":17,
        "duplicate":true,
        "ip":"fc08::2",
        "afi":"ipv6",
        "interfaceIndex":49,
        "interfaceName":"Ethernet4",
        "vrf":"default",
        "resolver":true,
        "active":true,
        "weight":1,
        "resolvedVia":117
      }
    ]
  }
}
PE3# show nexthop rib 145     
ID: 145 (zebra)
     RefCnt: 1
     Uptime: 00:03:27
     VRF: default(No AFI)
     Nexthop Count: 2
     Flags: 0x3
     Valid, Installed
     Depends: (142) (146)
        via 2064:100::1d (vrf default) (recursive), weight 1
           via fc06::2, Ethernet12 (vrf default), weight 1
           via fc08::2, Ethernet4 (vrf default), weight 1
        via 2064:200::1e (vrf default) (recursive), weight 1
           via fc06::2, Ethernet12 (vrf default) (dup), weight 1
           via fc08::2, Ethernet4 (vrf default) (dup), weight 1
PE3# 
PE3# 
PE3# 
PE3# 
PE3# show nexthop rib 142 json
{
  "142":{
    "type":"zebra",
    "refCount":2,
    "uptime":"00:03:35",
    "vrf":"default",
    "afi":"IPv6",
    "nexthopCount":1,
    "flags":16393,
    "valid":true,
    "installedFpmOnly":true,
    "recursive":true,
    "depends":[
      117
    ],
    "nexthops":[
      {
        "flags":5,
        "ip":"2064:100::1d",
        "afi":"ipv6",
        "vrf":"default",
        "active":true,
        "recursive":true,
        "weight":1
      },
      {
        "flags":3,
        "fib":true,
        "ip":"fc06::2",
        "afi":"ipv6",
        "interfaceIndex":51,
        "interfaceName":"Ethernet12",
        "vrf":"default",
        "resolver":true,
        "active":true,
        "weight":1,
        "resolvedVia":117
      },
      {
        "flags":3,
        "fib":true,
        "ip":"fc08::2",
        "afi":"ipv6",
        "interfaceIndex":49,
        "interfaceName":"Ethernet4",
        "vrf":"default",
        "resolver":true,
        "active":true,
        "weight":1,
        "resolvedVia":117
      }
    ],
    "dependents":[
      145
    ]
  }
}
PE3# show nexthop rib 142     
ID: 142 (zebra)
     RefCnt: 2
     Uptime: 00:03:37
     VRF: default(IPv6)
     Nexthop Count: 1
     Flags: 0x4009
     Valid, Installed (FPM only), Recursive
     Depends: (117)
        via 2064:100::1d (vrf default) (recursive), weight 1
           via fc06::2, Ethernet12 (vrf default), weight 1
           via fc08::2, Ethernet4 (vrf default), weight 1
     Dependents: (145)


From current output, this resolve via would help to handle PIC local case, a.k.a local link down case.

But it would not be able to handle PIC core case, which handles recursive routes withdrew case, which we need to handle  2064:100::1d withdrew case in global table, or PIC  edge case, which we need to handle 2064:100::1d withdrew case in VPN case.

mjstapp commented:
> @mjstapp Put your diff into our testbed. We could get resolvedVia in json, but I don't see resolve via in regular "show nexthop rib " output

[...]

> From current output, this resolve via would help to handle PIC local case, a.k.a local link down case.
> 

Yes, you can see in the diff where the show output is produced. I'll take a look at the nhg show output and see about adding the info there also.

> But it would not be able to handle PIC core case, which handles recursive routes withdrew case, which we need to handle 2064:100::1d withdrew case in global table, or PIC edge case, which we need to handle 2064:100::1d withdrew case in VPN case.

I'm not sure what "it" is there - from that blast of show output, I can't extract the problem?

eddieruan-alibaba commented:
@mjstapp Use this example 

NHG 131 (recursive singleton for 2064:200::1e, Depends: (109)):

The route to 2064:200::1e is installed with NHE 109 (that's exactly why 131 depends on 109 — a recursive NHG's depend is the resolving route's NHE). So:
via 2064:200::1e (recursive), weight 1 <- resolved_via = 0 (not shown)
via fc06::2, Ethernet12, weight 1 <- resolved_via = 109
via fc08::2, Ethernet4, weight 1 <- resolved_via = 109

NHG 136 (group of 131 + 137):

137 is the recursive singleton for 2064:100::1d. Its resolving route's nexthops are the identical pair (fc06::2/Ethernet12, fc08::2/Ethernet4), and since zebra shares NHEs by hash, that route's nhe_id is also 109.
136's nexthop list is a deep copy of its depends' nexthops (resolved_via is preserved by the copy — the PR adds it to nexthop_copy_no_recurse()), so:
via 2064:100::1d (recursive), weight 1 <- resolved_via = 0
via fc06::2, Ethernet12, weight 1 <- resolved_via = 109
via fc08::2, Ethernet4, weight 1 <- resolved_via = 109
via 2064:200::1e (recursive), weight 1 <- resolved_via = 0
via fc06::2, Ethernet12 (dup), weight 1 <- resolved_via = 109
via fc08::2, Ethernet4 (dup), weight 1 <- resolved_via = 109


When 2064:100::1d is withdrew,  we would like to make a quick fixup on NHG 136 to remove failed path for PIC core case.

But current resolve-via approach would not be able to help to trigger a fix up on NHG 136, right?

mjstapp commented:
the dplane sees nhg 131 (for example) go through. the dplane also sees any route that uses 131, and any nhg that includes 131. I'm not sure what you mean by "not be able"? if you want to, can't you maintain some data that will let you react if the route that controls nhg 131 is uninstalled?


> @mjstapp Use this example
> 
> NHG 131 (recursive singleton for 2064:200::1e, Depends: (109)):
> 
> The route to 2064:200::1e is installed with NHE 109 (that's exactly why 131 depends on 109 — a recursive NHG's depend is the resolving route's NHE). So: via 2064:200::1e (recursive), weight 1 <- resolved_via = 0 (not shown) via fc06::2, Ethernet12, weight 1 <- resolved_via = 109 via fc08::2, Ethernet4, weight 1 <- resolved_via = 109
> 
> NHG 136 (group of 131 + 137):
> 
> 137 is the recursive singleton for 2064:100::1d. Its resolving route's nexthops are the identical pair (fc06::2/Ethernet12, fc08::2/Ethernet4), and since zebra shares NHEs by hash, that route's nhe_id is also 109. 136's nexthop list is a deep copy of its depends' nexthops (resolved_via is preserved by the copy — the PR adds it to nexthop_copy_no_recurse()), so: via 2064:100::1d (recursive), weight 1 <- resolved_via = 0 via fc06::2, Ethernet12, weight 1 <- resolved_via = 109 via fc08::2, Ethernet4, weight 1 <- resolved_via = 109 via 2064:200::1e (recursive), weight 1 <- resolved_via = 0 via fc06::2, Ethernet12 (dup), weight 1 <- resolved_via = 109 via fc08::2, Ethernet4 (dup), weight 1 <- resolved_via = 109
> 
> When 2064:100::1d is withdrew, we would like to make a quick fixup on NHG 136 to remove failed path for PIC core case.
> 
> But current resolve-via approach would not be able to help to trigger a fix up on NHG 136, right?

eddieruan-alibaba commented:
NHG 131 is recursive NH, which would not be downloaded to dplane. 

In zebra code

https://github.com/FRRouting/frr/blob/master/zebra/zebra_nhg.c#L1284

we skip recursive NHG install to kernel.

Then we collect recursive's depend NHG in zebra_nhg_nhe2grp_internal()

https://github.com/FRRouting/frr/blob/master/zebra/zebra_nhg.c#L3418

Routes would not use 131, but use its depend which is 109. 

That's why we have the request to download recursive NH to dplane. 
https://github.com/FRRouting/frr/issues/22733

mjstapp commented:
I understand that the kernel install uses a flattened list - but we have the complete tree of nexthops in the context, copied with `nexthop_group_copy`. Are you saying something is missing from that?


> NHG 131 is recursive NH, which would not be downloaded to dplane.
> 
> In zebra code
> 
> https://github.com/FRRouting/frr/blob/master/zebra/zebra_nhg.c#L1284
> 
> we skip recursive NHG install to kernel.
> 
> Then we collect recursive's depend NHG in zebra_nhg_nhe2grp_internal()
> 
> https://github.com/FRRouting/frr/blob/master/zebra/zebra_nhg.c#L3418
> 
> Routes would not use 131, but use its depend which is 109.
> 
> That's why we have the request to download recursive NH to dplane. #22733

mjstapp commented:
an example:
```
C>* 4.4.4.0/24 (15) is directly connected, GRETA, weight 1, 02:41:23
L>* 4.4.4.1/32 (15) is directly connected, GRETA, weight 1, 02:41:23
S>* 5.5.5.5/32 [1/0] (25) via 4.4.4.44, GRETA, weight 1, 02:38:14
S>  6.6.6.6/32 [1/0] (29) via 5.5.5.5 (recursive), weight 1, 02:35:56
  *                         via 4.4.4.44, GRETA, weight 1, 02:35:56

```

so I have a recursive route 6.6.6.6, and it's dependent on 5.5.5.5. If 5.5.5.5 is withdrawn, you can react to that, if you want to?

eddieruan-alibaba commented:
Use your example

C>* 4.4.4.0/24 is directly connected, Ethernet4, weight 1, 00:00:55
L>* 4.4.4.0/32 is directly connected, Ethernet4, weight 1, 00:00:55
S>* 5.5.5.5/32 [1/0] via 4.4.4.44, Ethernet4, weight 1, 00:00:20
S>  6.6.6.6/32 [1/0] via 5.5.5.5 (recursive), weight 1, 00:00:03
  *                    via 4.4.4.44, Ethernet4, weight 1, 00:00:03
  
  zebra RE 6.6.6.6/32 points to recursive NHE 165. 
  
  PE3# show ip route 6.6.6.6 nexthop-group 
Routing entry for 6.6.6.6/32
  Known via "static", distance 1, metric 0, best
  Last update 00:00:35 ago
  Flags: Recursion Selected RR Distance Offloaded 
  Status: Installed 
  Nexthop Group ID: 165
  Installed Nexthop Group ID: 163
  Received Nexthop Group ID: 164
    5.5.5.5 (recursive), weight 1
  *   4.4.4.44, via Ethernet4, weight 1, res via 163

PE3# show nexthop rib 165
ID: 165 (zebra)
     RefCnt: 1
     Uptime: 00:00:47
     VRF: default(IPv4)
     Nexthop Count: 1
     Flags: 0x4009
     Valid, Installed (FPM only), Recursive
     Depends: (163)
        via 5.5.5.5 (vrf default) (recursive), weight 1
           via 4.4.4.44, Ethernet4 (vrf default), weight 1
           
When zebra decides to install it to kernel, zebra uses 

        zebra_nhg_install_kernel() — [zebra/zebra_nhg.c:3538: ](https://github.com/FRRouting/frr/blob/master/zebra/zebra_nhg.c#L3538)   
                                                                                                              
This function would trigger zebra_nhg_resolve() first                                               
    /* Resolve it first */
    nhe = zebra_nhg_resolve(nhe);
https://github.com/FRRouting/frr/blob/master/zebra/zebra_nhg.c#L3543
                                                                                    
zebra_nhg_resolve() — [zebra/zebra_nhg.c:212-221](https://github.com/FRRouting/frr/blob/master/zebra/zebra_nhg.c#L212C1-L221C2) 
peels recursive NHEs: while NEXTHOP_GROUP_RECURSIVE, replace nhe with its depend. 
So a recursive NHE itself never generates a dplane ctx — only its resolved depend does.

dplane context is created after calling 

enum zebra_dplane_result ret = dplane_nexthop_add(nhe);
https://github.com/FRRouting/frr/blob/master/zebra/zebra_nhg.c#L3575

which calls 
https://github.com/FRRouting/frr/blob/master/zebra/zebra_dplane.c#L5042
    ret = dplane_ctx_nexthop_init(ctx, op, nhe);

Since we have already peeled all recursive NHE, the NHE passed to  dplane_ctx_nexthop_init is recursive's depend NHE.
 
dplane_ctx_nexthop_init() — https://github.com/FRRouting/frr/blob/master/zebra/zebra_dplane.c#L4255

triggers deep copy from recursive NHE's depend NHE, which doesn't have recursive NH information anymore. 

nexthop_group_copy(&ctx->u.rinfo.nhe.ng, &nhe->nhg)


But in dplane_ctx_route_init()

	/* Copy nexthops; recursive info is included too */
	copy_nexthops(&(ctx->u.rinfo.zd_ng.nexthop),
		      re->nhe->nhg.nexthop, NULL);

https://github.com/FRRouting/frr/blob/master/zebra/zebra_dplane.c#L4054

a.k.a for DPLANE_OP_ROUTE_UPDATE, we have full nexthop tree which includes recursive NH information.

But NHG updates are from the following dplane apis for different ops, which we skip recursive NH's installation. 

/*
 * Enqueue a nexthop add for the dataplane.
 */
enum zebra_dplane_result dplane_nexthop_add(struct nhg_hash_entry *nhe)
{
	enum zebra_dplane_result ret = ZEBRA_DPLANE_REQUEST_FAILURE;

	if (nhe)
		ret = dplane_nexthop_update_internal(nhe, DPLANE_OP_NH_INSTALL);
	return ret;
}

/*
 * Enqueue a nexthop update for the dataplane.
 *
 * Might not need this func since zebra's nexthop objects should be immutable?
 */
enum zebra_dplane_result dplane_nexthop_update(struct nhg_hash_entry *nhe)
{
	enum zebra_dplane_result ret = ZEBRA_DPLANE_REQUEST_FAILURE;

	if (nhe)
		ret = dplane_nexthop_update_internal(nhe, DPLANE_OP_NH_UPDATE);
	return ret;
}

For PIC, we want to use nexthop events to build up NHG hierarchical dependencies, instead of routes event which would be prefix dependent.

eddieruan-alibaba commented:
The other question is how we plan to use resolve via for convergence. 

In your example, 5.5.5.5 is via 4.4.4.44, 6.6.6.6 via 5.5.5.5. We have

PE3# show ip route 6.6.6.6 next
Routing entry for 6.6.6.6/32
  Known via "static", distance 1, metric 0, best
  Last update 00:39:38 ago
  Flags: Recursion Selected RR Distance Offloaded 
  Status: Installed 
  Nexthop Group ID: 165
  Installed Nexthop Group ID: 163
  Received Nexthop Group ID: 164
    5.5.5.5 (recursive), weight 1
  *   4.4.4.44, via Ethernet4, weight 1, res via 163
  
PE3# show ip route
Codes: K - kernel route, C - connected, L - local, S - static,
       R - RIP, O - OSPF, I - IS-IS, B - BGP, E - EIGRP, N - NHRP,
       T - Table, v - VNC, V - VNC-Direct, A - Babel, F - PBR,
       f - OpenFabric, t - Table-Direct,
       > - selected route, * - FIB route, q - queued, r - rejected, b - backup
       t - trapped, o - offload failure

Since 4.4.4.44/32 is ARP resolved prefix for Ethernet 4 with interface prefix 4.4.4.0/24. Zebra's routing table doesn't have this route.

IPv4 unicast VRF default:
C>* 4.4.4.0/24 is directly connected, Ethernet4, weight 1, 00:41:28
L>* 4.4.4.0/32 is directly connected, Ethernet4, weight 1, 00:41:28
S>* 5.5.5.5/32 [1/0] via 4.4.4.44, Ethernet4, weight 1, 00:40:53
S>  6.6.6.6/32 [1/0] via 5.5.5.5 (recursive), weight 1, 00:40:36
                      via 4.4.4.44, Ethernet4, weight 1, 00:40:36

Therefore we need to build an relationship for 
4.4.4.0/24 with nexthop 4.4.4.44 in dplane to track the binding relationship, which would be essentially similar to NHT created once in zebra.

PE3# show ip nht 
VRF default:
 Resolve via default: on
4.4.4.44
 resolved via connected, prefix 4.4.4.0/24
 is directly connected, Ethernet4 (vrf default), weight 1
 Client list: static(fd 72)
5.5.5.5
 resolved via static, prefix 5.5.5.5/32
 via 4.4.4.44, Ethernet4 (vrf default), weight 1
 Client list: static(fd 72)


Do you have any suggestions on how to build up this binding relationship without LPM gets involved. I try not to build another LPM in dplane or SONiC side which might be too expensive.

Commit:
WIP adding resolving prefix in addition to the resolving NHG ID.
Trying this out for now, adding cells to the nexthop struct;
that may not be what we want to do permanently though.

Signed-off-by: Mark Stapp <mjs@cisco.com>

riw777 reviewed (approved):
looks good

eddieruan-alibaba commented:
admin@PE3:~$ vtysh -c "show run" | grep "ipv6 route"
ipv6 route 1::/64 fc06::2 Ethernet12
ipv6 route 1::/64 fc08::2 Ethernet4
ipv6 route 3::/64 fc06::2 Ethernet12
ipv6 route 2::2/128 1::1
ipv6 route 2::2/128 3::3
admin@PE3:~$ 
admin@PE3:~$ 
admin@PE3:~$ 
admin@PE3:~$ vtysh -c "show ipv6 route 2::2 next"
Routing entry for 2::2/128
  Known via "static", distance 1, metric 0, best
  Last update 00:00:37 ago
  Flags: Recursion Selected RR Distance Offloaded 
  Status: Installed 
  Nexthop Group ID: 1129
  Received Nexthop Group ID: 1129
    1::1 (recursive), weight 1
  *   fc06::2, via Ethernet12, weight 1, res via 1::/64 (129)
  *   fc08::2, via Ethernet4, weight 1, res via 1::/64 (129)
    3::3 (recursive), weight 1
      fc06::2, via Ethernet12 (duplicate nexthop removed), weight 1, res via 3::/64 (128)


@mjstapp 
Two questions / suggestions
1. 
 use above static routes formed recursive case as an example.

Should be make resolve via prefix / NHG as a attribute for recursive NH?

a.k.a show 

    1::1 (recursive), weight 1 res via 1::/64 (129)
    3::3 (recursive), weight 1 res via 3::/64 (128)
    
   instead of currently keep them as  resolved NH's attribute. 

From convergence point of view, if 1::/64  route gets deleted.  I would use 1::/64  to find out recursive NHG for 1::1 and  trigger backwalk to fix up the ECMP for NH 1::1 and 3::3. 

2. For  NH  fc06::2 via Ethernet12, would you be able to provide  that it is resolve via its interface prefix fc06::/120

C>* fc06::/120 is directly connected, Ethernet12, weight 1, 10:49:39
L>* fc06::1/128 is directly connected, Ethernet12, weight 1, 10:49:39

So I could use fc06::/120 rotue delete event to trigger backwalk to fix up all ECMP using  this NH. 

As we discussed last time, SONiC dplane will maintain dplane NHG based on route events and track resolved via route to resolved dplane NHG ids mapping.

mjstapp commented:
[...]

> @mjstapp Two questions / suggestions 1. use above static routes formed recursive case as an example.
> 
> Should be make resolve via prefix / NHG as a attribute for recursive NH?
> 
> a.k.a show
> 
> ```
> 1::1 (recursive), weight 1 res via 1::/64 (129)
> 3::3 (recursive), weight 1 res via 3::/64 (128)
> ```
> 
> instead of currently keep them as resolved NH's attribute.
> 

yes, I think we can do that - I'll push an update we can look at

[...]
> 
>     2. For  NH  fc06::2 via Ethernet12, would you be able to provide  that it is resolve via its interface prefix fc06::/120
> 

so that means you'd like the "resolving prefix" info even for non-recursive routes, for "regular" routes that use connected prefixes directly? I think that's doable too - I will work on an update.

mjstapp commented:
A question, for @eddieruan-alibaba  Eddie/sonic :
there's a second "resolving" path that zebra uses, using a SR-TE "color" and matching policy object. that path is in the block beginning with  `if (nexthop->srte_color)  { ...`
That code path doesn't use the same "resolving prefix" logic: is there something that you would need added in that path?

mjstapp commented:
Here's an example with a couple of the changes (I think) - how does this info look?

(I've added this to the normal show route output for now, but it does clutter things up a bit, maybe we should only display it in "detail" mode? or json?)

```
C>* 2.2.2.0/24 is directly connected, ANNIE, weight 1, 00:00:47
L>* 2.2.2.1/32 is directly connected, ANNIE, weight 1, 00:00:47
C>* 3.3.3.0/24 is directly connected, BETTY, weight 1, 00:00:47
L>* 3.3.3.1/32 is directly connected, BETTY, weight 1, 00:00:47
C>* 4.4.4.0/24 is directly connected, GRETA, weight 1, 00:00:47
L>* 4.4.4.1/32 is directly connected, GRETA, weight 1, 00:00:47
S>* 5.5.5.5/32 [1/0] via 2.2.2.22, ANNIE, label 222, weight 1, res via 2.2.2.0/24 (13), 00:00:16
  *                  via 3.3.3.33, BETTY, label 333, weight 1, res via 3.3.3.0/24 (14), 00:00:16
S>  6.6.6.6/32 [1/0] via 5.5.5.5 (recursive), label 666, weight 1, res via 5.5.5.5/32 (30), 00:00:05
  *                    via 2.2.2.22, ANNIE, label 222/666, weight 1, 00:00:05
  *                    via 3.3.3.33, BETTY, label 333/666, weight 1, 00:00:05
S>  7.7.7.7/32 [1/0] via 6.6.6.6 (recursive), label 777, weight 1, res via 6.6.6.6/32 (35), 00:00:02
  *                    via 2.2.2.22, ANNIE, label 222/666/777, weight 1, 00:00:02
  *                    via 3.3.3.33, BETTY, label 333/666/777, weight 1, 00:00:02

```

eddieruan-alibaba commented:
> Here's an example with a couple of the changes (I think) - how does this info look?
> 
> (I've added this to the normal show route output for now, but it does clutter things up a bit, maybe we should only display it in "detail" mode? or json?)
> 
> ```
> C>* 2.2.2.0/24 is directly connected, ANNIE, weight 1, 00:00:47
> L>* 2.2.2.1/32 is directly connected, ANNIE, weight 1, 00:00:47
> C>* 3.3.3.0/24 is directly connected, BETTY, weight 1, 00:00:47
> L>* 3.3.3.1/32 is directly connected, BETTY, weight 1, 00:00:47
> C>* 4.4.4.0/24 is directly connected, GRETA, weight 1, 00:00:47
> L>* 4.4.4.1/32 is directly connected, GRETA, weight 1, 00:00:47
> S>* 5.5.5.5/32 [1/0] via 2.2.2.22, ANNIE, label 222, weight 1, res via 2.2.2.0/24 (13), 00:00:16
>   *                  via 3.3.3.33, BETTY, label 333, weight 1, res via 3.3.3.0/24 (14), 00:00:16
> S>  6.6.6.6/32 [1/0] via 5.5.5.5 (recursive), label 666, weight 1, res via 5.5.5.5/32 (30), 00:00:05
>   *                    via 2.2.2.22, ANNIE, label 222/666, weight 1, 00:00:05
>   *                    via 3.3.3.33, BETTY, label 333/666, weight 1, 00:00:05
> S>  7.7.7.7/32 [1/0] via 6.6.6.6 (recursive), label 777, weight 1, res via 6.6.6.6/32 (35), 00:00:02
>   *                    via 2.2.2.22, ANNIE, label 222/666/777, weight 1, 00:00:02
>   *                    via 3.3.3.33, BETTY, label 333/666/777, weight 1, 00:00:02
> ```

Looks great

eddieruan-alibaba commented:
> A question, for @eddieruan-alibaba Eddie/sonic : there's a second "resolving" path that zebra uses, using a SR-TE "color" and matching policy object. that path is in the block beginning with `if (nexthop->srte_color) { ...` That code path doesn't use the same "resolving prefix" logic: is there something that you would need added in that path?

You bought a very good question. 

Both SR-TE, RSVP-TE would have similar situation. All of them are logically treated as point to point tunnel interfaces. 

For SR-TE case, we will keep NH up if there is no BFD protecting it, otherwise, we rely on BFD to toggle NH up and down status. Routes are bind with these NHs via nexthop + color via policy and nexthop could be "::" or "0.0.0.0" for color only case. 

 For my previous second ask, " For  NH  fc06::2 via Ethernet12, would you be able to provide  that it is resolve via its interface prefix fc06::/120", it only solves point to multi-point type of interface's NH 's status tracking, which could be used in PIC local convergence case. But we don't have a good solution for ECMP via mutlple SR-TE paths case' quick fix up.

eddieruan-alibaba commented:
> > A question, for @eddieruan-alibaba Eddie/sonic : there's a second "resolving" path that zebra uses, using a SR-TE "color" and matching policy object. that path is in the block beginning with `if (nexthop->srte_color) { ...` That code path doesn't use the same "resolving prefix" logic: is there something that you would need added in that path?
> 
> You bought a very good question.
> 
> Both SR-TE, RSVP-TE would have similar situation. All of them are logically treated as point to point tunnel interfaces.
> 
> For SR-TE case, we will keep NH up if there is no BFD protecting it, otherwise, we rely on BFD to toggle NH up and down status. Routes are bind with these NHs via nexthop + color via policy and nexthop could be "::" or "0.0.0.0" for color only case.
> 
> For my previous second ask, " For NH fc06::2 via Ethernet12, would you be able to provide that it is resolve via its interface prefix fc06::/120", it only solves point to multi-point type of interface's NH 's status tracking, which could be used in PIC local convergence case. But we don't have a good solution for ECMP via mutlple SR-TE paths case' quick fix up.

@mjstapp  discuss with my team. We could drop this fast convergence  handling until the use cases are more clear. 

In Alibaba's deployment, currently, the number of routes over a SRv6 policy is not high. We could tolerate with regular convergence. So we could park it unless you see some other needed use cases.

eddieruan-alibaba commented:
Here is show rib nexthop output with your changes


PE3# show nexthop rib 2138
ID: 2138 (zebra)
     RefCnt: 2
     Uptime: 00:00:58
     VRF: default(No AFI)
     Nexthop Count: 2
     Flags: 0x3
     Valid, Installed
     Depends: (2134) (2139)
        via 2064:100::1d (vrf default) (recursive), weight 1, res via 2064:100::1d/128 (124)
           via fc06::2, Ethernet12 (vrf default), weight 1
           via fc08::2, Ethernet4 (vrf default), weight 1
        via 2064:200::1e (vrf default) (recursive), weight 1, res via 2064:200::1e/128 (124)
           via fc06::2, Ethernet12 (vrf default) (dup), weight 1
           via fc08::2, Ethernet4 (vrf default) (dup), weight 1
PE3# show nexthop rib 124
ID: 124 (zebra)
     RefCnt: 41
     Uptime: 00:00:49
     VRF: default(No AFI)
     Nexthop Count: 2
     Flags: 0x3
     Valid, Installed
     Depends: (117) (123)
        via fc06::2, Ethernet12 (vrf default), weight 1, res via fc06::/120 (41)
        via fc08::2, Ethernet4 (vrf default), weight 1, res via fc08::/120 (61)
     Dependents: (2134) (2139) (2141)

eddieruan-alibaba commented:
Here is SONIC fmp output to show dplane NHG, its relateship with zebra NHG and resolved prefix.

PE3# show fpm nhg-fib 
Dplane NHG 1 (L-C) flags 0x0000 refcount 3 hash 0xefc769fd586109a2
  rib nhg ids: 15 9
Dplane NHG 2 (L-C) flags 0x0000 refcount 2 hash 0x19c2ce58c571417a
  rib nhg ids: 32 33
Dplane NHG 3 (L-C) flags 0x0000 refcount 2 hash 0x31029f077454ae31
  rib nhg ids: 41
Dplane NHG 4 (L-C) flags 0x0000 refcount 2 hash 0x0c61b9eca39f4071
  rib nhg ids: 157
Dplane NHG 5 (L-C) flags 0x0000 refcount 5 hash 0xf8b417b4d1ab8a85
  rib nhg ids: 52 53
Dplane NHG 6 (L-C) flags 0x0000 refcount 7 hash 0x8d87ecf3131484c4
  resolved via fc08::/120 vrf 0 rib nhg 61
  rib nhg ids: 117
Dplane NHG 7 (L-C) flags 0x0000 refcount 7 hash 0x6671082c87017642
  resolved via fc06::/120 vrf 0 rib nhg 41
  rib nhg ids: 123
Dplane NHG 8 (L-A) flags 0x0000 refcount 15 hash 0x60602799e38c368f
  children: 7(w1) 6(w1)
  rib nhg ids: 124
Dplane NHG 9 (L-A) flags 0x1000 refcount 10 hash 0xfe4ccd0cce40bd18
  children: 10(w1) 11(w1)
  rib nhg ids: 136
Dplane NHG 10 (L-C) flags 0x1000 refcount 1 hash 0x05361d4e138589f0
  resolved via 2064:200::1e/128 vrf 0 rib nhg 124
Dplane NHG 11 (L-C) flags 0x1000 refcount 1 hash 0xa52519104805a35e
  resolved via 2064:100::1d/128 vrf 0 rib nhg 124
Dplane NHG 12 (L-C) flags 0x0000 refcount 10 hash 0x8108501e73835759
  resolved via 10.10.246.0/24 vrf 76 rib nhg 52
  rib nhg ids: 145
Dplane NHG 13 (L-B) flags 0x0008 refcount 1 hash 0x4013656f2339679a
  resolved via 2064:100::1d/128 vrf 0 rib nhg 124
  children: 7(w1) 6(w1)
Dplane NHG 14 (L-B) flags 0x0008 refcount 1 hash 0x132dcdd5d61b94cc
  resolved via 1::1/128 vrf 0 rib nhg 2138
  children: 7(w1) 6(w1)
Dplane NHG 15 (L-B) flags 0x0008 refcount 1 hash 0x969310bfb78a6e00
  resolved via 2064:200::1e/128 vrf 0 rib nhg 124
  children: 7(w1) 6(w1)
Dplane NHG 16 (L-A) flags 0x0000 refcount 1 hash 0xab707974bfe6fee5
  children: 13(w1) 15(w1)
  rib nhg ids: 2138
Dplane NHG 17 (L-A) flags 0x0000 refcount 1 hash 0x6ac46d058dd2559d
  children: 14(w1)
  rib nhg ids: 124
PE3# 
PE3# 
PE3# 
PE3# show fpm nhg-fib resolved-via 
Resolving prefix                         VRF  Resolved by  Dplane NHG ids
fc06::/120                               0    rib 41       7
fc08::/120                               0    rib 61       6
1::1/128                                 0    rib 2138     14
2064:100::1d/128                         0    rib 124      11 13
2064:200::1e/128                         0    rib 124      10 15
10.10.246.0/24                           76   rib 52       12

eddieruan-alibaba reviewed (approved):
Verified with SONiC.

GaladrielZhao reviewed (approved):
Thanks, looks good to me.

mjstapp commented:
pushed a round of fixes, including removing the "resolving" info from the default cli show output  -  that breaks some topotests (at least) that still depend on the text output rather than json. removed the 'don't merge' label now, should be ready for review.

mjstapp commented:
wow - there are topotests that fail if new json keys are added. I've added the new "resolved via" keys to the test that is failed here. I will fix the rest of the tests so that this doesn't happen again.
